Add support for FrSky OSD

fiam opened the pull request:
- Add displayWriteFontCharacter() for font writing, removing all max7456
specific code.
- Add displayIsReady() for asynchronous display initialization
- Add displayBeginTransaction()/displayCommitTransaction() for display
transactions, which allow performing complex drawing operations without
flickering
- Add displayGetCanvas(), which retrieves the canvas associated with a
display (if it has it)
- Add canvas implementation for pixel based access for a display
- Add FrSkyOSD driver and displayPort driver
- Enable FrSkyOSD driver for targets with flash > 256
- Rename max7456_symbols.h to osd_symbols.h

mikeller commented:
@fiam: This looks interesting. Can you please provide a bit more detail on what it does:
- how can this be used;
- what are the differences compared to the existing displayPort implementation;
- what devices support it;
- where can the specification of the protocol that can be used by third parties to implement this on their devices be found.

Maybe adding a page to `/docs/` that contains these points and some user facing instructions for set up would be helpful as well.

fiam commented:
@mikeller Sure! This is just the OSD driver and some facilities for pixel based drawing. With this code, it kinda has the same level of functionality than a MAX7456 (but with way less current consumption and smaller footprint). Once this is merged, I'd like to discuss with you guys which OSD elements would make sense to implement as pixel based in BF and submit a PR for them.

The displayPort changes just add a bit of functionality for asynchronous display initialisation (these devices will be available both as SMD modules for FC integration and external boards over UART), as well as for transactions (these allow performing intensive draw operations without flickering).

The canvas layer is entirely new and it consists of pixel/vector drawing operations as well as some functions for managing transformations and the drawing contexts. The reason I went with a separate vtable is because a) displayPort felt big enough already and b) I didn't want to bloat all displayPort drivers with these pixel based operations function pointers when we only have a device supporting them right now. Of course, it will be nice that as more pixel based OSDs hit the market, we build upon and expand this canvas API as needed instead of having a separate interface for each device.

Regarding the available devices, there are some FCs from FrSky and other manufacturers (which I'm not sure I'm allowed to disclose)coming out soon with this OSD instead of a MAX7456. There's also a separate small OSD board that connects to the FC via UART in the works.

Regarding the protocol specification, FrSky is deciding how to publish it. I've recommended them to make it public, but that's not my call to make. I'll let guys know as soon as I have more news on this front.

mikeller commented:
@fiam:

> Regarding the protocol specification, FrSky is deciding how to publish it. I've recommended them to make it public, but that's not my call to make. I'll let guys know as soon as I have more news on this front.

I think what we want to avoid from Betaflight's side of things is yet another situation where code is 'contributed' to the GPLv3 project, but a manufacturer tries to own the protocol specification. This has only given us grief and led to bad user experience in the past, as evidenced with problems with TBS SmartAudio.
So as an additional motivator for FrSky to publicise this, I do not think that Betaflight will want to add support for this unless there is a publicly available protocol specification for the hardware.

fiam commented:
Addressing some comments from the INAV PR that I wasn't able to reply due to time constraints and might be more relevant on this PR at this point.

@jflyper This doesn't use MSP_DISPLAYPORT, it has its own protocol. I'm hoping it will be made publicly available soon, but I can't give you an exact date. Some rough specs: 360x288 (same as MAX7456), compatible with MAX7456 font files, support for individual pixel access and 3 colors (black, white and grey). The API supports stroking and filling some shapes as well as drawing arbitrary lines. There's also a CTM which performs affine transforms when drawing (handy for implementing stuff like the AHI).

@hydra The reason the path based filling API is not present at the moment is because it's slow, significantly slower than drawing specific shapes like triangles, rects or ellipses. I didn't want to encourage developers to use an API like that, because you can put stuff on the screen (or update it faster) using the lower level APIs. The existing API should cover 90%+ of the use cases while allowing more frames per second to be drawn. If once you get a chance to play with this you feel the existing API is limiting the kind of visual elements that you can build, please let me know and we'll see what can be done about it.

hydra commented:
@fiam thabks for the reply!  Long time no chat.

Yes, would love to have a play with one.  Any chance you can organize samples for interested developers?

fiam commented:
@hydra My pleasure :-). I've just asked FrSky if they can provide some samples for BF developers, but it's late in CN right now so I wouldn't expect a reply today. I'll let you guys know when I have more information.

mikeller commented:
The protocol specification for this has been published now: https://github.com/FrSkyRC/PixelOSD/blob/master/doc/OSD.md

mikeller commented on src/main/common/utils.h:
```diff
@@ -118,3 +118,5 @@ void * memcpy_fn ( void * destination, const void * source, size_t num ) asm("me
 #define FALLTHROUGH do {} while(0)
 #endif
+
+#define ALIGNED(x) __attribute__ ((aligned(x)))
```
Is this actually used anywhere?

fiam replied:
My mistake. I left it over while porting the driver from INAV.

mikeller commented on src/main/common/uvarint.c:
```diff
@@ -0,0 +1,72 @@
+{
+    unsigned ii = 0;
+    while (val > 0x80)
+    {
```
Should go onto the same line.

mikeller commented on src/main/io/displayport_frsky_osd.c:
```diff
@@ -0,0 +1,499 @@
+/*
+ * This file is part of Cleanflight.
```
This is an outdated header - copy the current one from any other source file.

mikeller commented on src/main/io/displayport_frsky_osd.h:
```diff
@@ -0,0 +1,7 @@
+#pragma once
```
This is missing the header.

mikeller commented on src/main/io/displayport_frsky_osd.c:
```diff
@@ -0,0 +1,499 @@
+{
+    UNUSED(instance);
+
+    frskyOSDWriteFontCharacter(addr, chr);
```
This seems to be discarding the output of `frskyOSDWriteFontCharacter()`.

mikeller commented on src/main/io/displayport_max7456.c:
```diff
@@ -152,6 +153,16 @@ static bool layerCopy(displayPort_t *displayPort, displayPortLayer_e destLayer,
     return max7456LayerCopy(destLayer, sourceLayer);
 }
 
+static int writeFontCharacter(displayPort_t *instance, uint16_t addr, const osdCharacter_t *chr)
```
Why the `int` return value here and not bool?

fiam replied:
The was to let the drivers return the number of bytes used, but it turns out that's not that useful right now. I'll change it to bool.

mikeller commented on src/main/io/frsky_osd.c:
```diff
@@ -0,0 +1,927 @@
+    RECV_STATE_DONE,
+} frskyOSDRecvState_e;
+
+typedef struct frskyOSDInfoResponse_s {
```
Commonly used form for camelCase would be `frskyOsdInfoResponse_s` (to avoid awkward capitalisations in situations like `oSDInit`).

fiam replied:
I'll do a global `s/frskyOSD/frskyOsd/g`

mikeller commented on src/main/io/frsky_osd.c:
```diff
@@ -0,0 +1,927 @@
+static bool frskyOSDHandleCommand(osdCommand_e cmd, const void *payload, size_t size)
+{
+    switch (cmd) {
+        case OSD_CMD_RESPONSE_ERROR:
```
`switch` and `case` are commonly indented at the same level in Betaflight.

mikeller commented on src/main/io/serial.h:
```diff
@@ -49,6 +49,7 @@ typedef enum {
     FUNCTION_VTX_TRAMP           = (1 << 13), // 8192
     FUNCTION_RCDEVICE            = (1 << 14), // 16384
     FUNCTION_LIDAR_TF            = (1 << 15), // 32768
+    FUNCTION_FRSKY_OSD           = (1 << 20), // 1048576
```
Why not 16?

fiam replied:
[16, 19] are already used in INAV. I thought that having the same FUNCTION id could be preferable in both firmwares, but using different ones is not a big problem either.

mikeller replied:
The values < 16 are not in synch between iNav and Betaflight already, so I don't think there is not much value in having this one in synch.

mikeller commented on src/main/msp/msp.c:
```diff
@@ -3280,15 +3281,39 @@ static mspResult_e mspCommonProcessInCommand(mspDescriptor_t srcDesc, uint8_t cm
 
     case MSP_OSD_CHAR_WRITE:
-#ifdef USE_MAX7456
+#if defined(USE_OSD)
```
lol, this is all already inside a giant `#if defined(USE_OSD)`: https://github.com/betaflight/betaflight/pull/9127/files#diff-eefd576c86ccfa8b6cdf98081806a7a8R3195

mikeller commented on src/main/msp/msp.c:
```diff
@@ -3280,15 +3281,39 @@ static mspResult_e mspCommonProcessInCommand(mspDescriptor_t srcDesc, uint8_t cm
-            const uint8_t addr = sbufReadU8(src);
-            for (int i = 0; i < 54; i++) {
-                font_data[i] = sbufReadU8(src);
+            osdCharacter_t chr;
```
Has this been tested and confirmed to work with a MAX7456?

fiam replied:
The MAX7456 driver continues to read only 54 bytes per character, it's not affected by this change.

mikeller commented on src/main/osd/osd.c:
```diff
@@ -386,15 +377,34 @@ void osdInit(displayPort_t *osdDisplayPortToUse)
+        return;
+    }
+
+    STATIC_ASSERT(OSD_POS_MAX == OSD_POS(31,31), OSD_POS_MAX_incorrect);
```
I think `STATIC_ASSERT()` is actually static these days, so this can be moved outside of a function body.

jflyper commented:
Did we talk about portableness of `packed` structures, especially when they are used over the wire yet?

jflyper commented:
@mikeller 
> @jflyper: You are working on something displayport related as well at the moment. Can you have a look at this to check if there is something that won't work with what you are doing?

`writeString` and `writeChar` in `io/displayport_frsky_osd.c` needs a signature change to accommodate the attribute rich osd support, but that's about it.

I have an F7 FC connected to both the frsky and an attribute rich osd, and I can switch between them without a problem by manipulating value of `osd_displayport_device`.

mikeller commented on src/main/common/uvarint.c:
```diff
@@ -0,0 +1,71 @@
+{
+    unsigned s = 0;
+    *val = 0;
+    for (size_t ii = 0; ii < size; ii++)
```
This file still has some `{`s that are not K&R.
